### ISSUE TITLE
[YSQL] reduce cost of un-prepared statements involving simple joins

### ISSUE THREAD

**Issue by kmuthukk** [yugabyte/yugabyte-db#1391]
For simple primary key based lookups, whether we use:
```
postgres=# PREPARE stmt AS select * from emp where eid=100;
postgres=# EXECUTE stmt;
```
or, un-prepared format:
```
postgres=# select * from emp where eid=100;
```
The timing different is not much. The execute step is about ~1.5ms for both and difference is not noticeable.

But for statements that involve even a simple join, the un-prepared version is about 9-10ms (instead of the 1.5-2ms range), suggesting that the query compilation time for some reason is a lot more.

For example, if the simple join statement is prepared upfront:

```
postgres=# PREPARE mystmt AS select eid, ename, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
```
then the execute phase is very quick:
```
postgres=# EXECUTE mystmt;
 eid |  ename   | salary | dept_name
-----+----------+--------+-----------
 100 | name-100 |  50000 | dname-100
(1 row)

Time: 2.330 ms
```

But the un-prepared version takes about 10-11ms.
```
postgres=# select eid, ename, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid |  ename   | salary | dept_name
-----+----------+--------+-----------
 100 | name-100 |  50000 | dname-100
(1 row)

Time: 11.120 ms
```

**Comment by kmuthukk** [comment]
@ddorian -- can you confirm if this still an issue?

**Comment by ddorian** [comment]
@kmuthukk yes it's still an issue.

1. Prepare schema
```
postgres=# \timing
postgres=# CREATE TABLE emp (eid INT primary key, salary bigint, dept_id bigint);
postgres=# create table dept(dept_id bigint primary key, dept_name text);
postgres=# PREPARE mystmt AS select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
PREPARE
Time: 88.284 ms
```
2. Time prepared statement:
```
postgres=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 46.432 ms
postgres=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 1.613 ms
postgres=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 1.433 ms
```
**Note the first run taking 46.432 ms.** Seems a lot too.

3. Time normal statement:
```
postgres=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 7.520 ms
postgres=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 7.885 ms
postgres=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 8.082 ms
```

**Comment by m-iancu** [comment]
@ddorian Can you confirm if this is still an issue today? Also to test let's run the simple select 6-7 times.

**Comment by ddorian** [comment]
@m-iancu  Looks fixed to me: (tested on 2.7.1.1)

```
yugabyte=# \timing
yugabyte=# CREATE TABLE emp (eid INT primary key, salary bigint, dept_id bigint);
yugabyte=# create table dept(dept_id bigint primary key, dept_name text);
yugabyte=# PREPARE mystmt AS select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
```

```
yugabyte=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 36.540 ms
yugabyte=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.220 ms
yugabyte=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.004 ms
yugabyte=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 1.974 ms
yugabyte=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.022 ms
yugabyte=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 1.994 ms
yugabyte=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 1.938 ms
yugabyte=# execute mystmt;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 1.882 ms
```
Simple join query:
```
yugabyte=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.356 ms
yugabyte=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.440 ms
yugabyte=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.296 ms
yugabyte=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.303 ms
yugabyte=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.260 ms
yugabyte=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.331 ms
yugabyte=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.314 ms
yugabyte=# select eid, salary, dept_name from emp, dept where eid=100 and emp.dept_id = dept.dept_id;
 eid | salary | dept_name 
-----+--------+-----------
(0 rows)

Time: 2.147 ms

```

**Comment by m-iancu** [comment]
Thanks for testing @ddorian, closing as it looks fixed.
